torch/onnx/_internal/exporter.py

```diff
@@ -410,43 +336,167 @@ def save(
+            # ONNX can't represent collection types (e.g., dictionary, tuple of tuple of
+            # tensor, etc), we flatten the collection and register each element as output.
+            self._output_adapter.append_step(io_adapter.FlattenOutputStep())
+
+        # Export TorchScript graph to ONNX ModelProto.
+        onnx_model = onnxscript_graph.to_model_proto(options.opset_version)
+        return torch.onnx.ExportOutput(
+            onnx_model, self._input_adapter, self._output_adapter
+        )
+
+    def adapt_input(
+        self,
+        adapt_step_cls: Type[io_adapter.InputAdaptStep],
+        model_args: Sequence[Any],
+        model_kwargs: Mapping[str, Any],
+        step_init_args: Optional[Sequence[Any]] = None,
+    ) -> Tuple[Sequence[Any], Mapping[str, Any]]:
+        """Apply an input adapt step to the model args and kwargs.
+
+        An input adapt step object is initialized, applied and recorded as part of
+        ``self._input_adapter`.
+
+        Args:
+            adapt_step_cls: The input adapt step class.
+            model_args: The model args.
+            model_kwargs: The model kwargs.
+            step_init_args: The input adapt step initialization arguments.
+
+        Returns:
+            The adapted model args and kwargs.
+        """
+        step_init_args = step_init_args or ()
+        adapt_step = adapt_step_cls(*step_init_args)
+        self._input_adapter.append_step(adapt_step)
+        return adapt_step.apply(model_args, model_kwargs)
+
+    def adapt_output(
+        self,
+        adapt_step_cls: Type[io_adapter.OutputAdaptStep],
+        model_outputs: Any,
+        step_init_args: Optional[Sequence[Any]] = None,
+    ) -> Any:
+        """Apply an output adapt step to the model outputs.
+
+        An output adapt step object is initialized, applied and recorded as part of
+        ``self._output_adapter`.
+
+        Args:
+            adapt_step_cls: The output adapt step class.
+            model_outputs: The model outputs.
+            step_init_args: The input adapt step initialization arguments.
+
+        Returns:
+            The adapted model outputs.
+        """
+        step_init_args = step_init_args or ()
+        adapt_step = adapt_step_cls(*step_init_args)
+        self._output_adapter.append_step(adapt_step)
+        return adapt_step.apply(model_outputs)
+
+    @abc.abstractmethod
+    def generate_fx(
+        self,
+        options: _ResolvedExportOptions,
+        model: Union[torch.nn.Module, Callable],
+        model_args: Sequence[Any],
+        model_kwargs: Mapping[str, Any],
+    ) -> Tuple[
+        torch.fx.GraphModule, Tuple[Any]
+    ]:  # By design, only torch.fx.GraphModule is needed
+        # But FXSymbolicTracer modifies model data, which will be needed
+        # to produce the ONNX proto in the next layer.
+        # TODO: Refactor after https://github.com/pytorch/pytorch/pull/98421
```

Comment:
This todo is outdated?

Reply:
Great catch!!!!!!!!
Actually, I forgot to refactor after you merged the io adapter stuff.
Now `generate_fx` does not need to return `Tuple[Any]` as we can reconstruct it through `fx_tracer` API

Reply:
Something is breaking, need to debug later. Will leave this TODO and refactor on a separate PR

Reply:
`adapt_input` and `adapt_output` can probably be removed and only `self.input_adapter.append_step` should be needed because the adapter input doesnt need to be returned anyway.

The only place this seems to fail is at `FXSymbolicTracer.generate_fx` in which

```python
    def generate_fx(
        self,
        options: exporter._ResolvedExportOptions,
        model: Union[torch.nn.Module, Callable],
        model_args: Sequence[Any],
        model_kwargs: Mapping[str, Any],
    ) -> Tuple[torch.fx.GraphModule, Tuple[Any]]:
        graph_module, bound_args = self._trace_into_fx_graph_via_fx_symbolic_trace(
            model, model_args, model_kwargs
        )
(...)
        replaced_attrs = replace_get_attr_with_placeholder_pass.replaced_attrs
(...)
        return graph_module, (*bound_args, *replaced_attrs)  # type: ignore[return-value]
```

`replaced_attrs` is not an input adapt step, but seems to change input, right? This is probably a sign that we need to figure out the passes interface ASAP to prevent this kind of hacking